rust/libnewsboat-ffi/src/history.rs

```diff
@@ -0,0 +1,121 @@
+use abort_on_panic;
+use libc::{c_char, c_void};
+use libnewsboat::history::History;
+use std::ffi::{CStr, CString};
+use std::mem;
+
+#[no_mangle]
+pub extern "C" fn rs_history_new() -> *mut c_void {
+    abort_on_panic(|| Box::into_raw(Box::new(History::new())) as *mut c_void)
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_history_free(hst: *mut c_void) {
+    abort_on_panic(|| {
+        if hst.is_null() {
+            return;
+        }
+        Box::from_raw(hst as *mut History);
+    })
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_history_add_line(hst: *mut c_void, line: *const c_char) {
+    abort_on_panic(|| {
+        let mut hst = {
+            assert!(!hst.is_null());
+            Box::from_raw(hst as *mut History)
+        };
+        let line = {
+            assert!(!line.is_null());
+            CStr::from_ptr(line)
+        }
+        .to_string_lossy()
+        .into_owned();
+
+        hst.add_line(line);
+
+        // Do not deallocate the object - C still has a pointer to it
+        mem::forget(hst);
+    })
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_history_prev(hst: *mut c_void) -> *mut c_char {
+    abort_on_panic(|| {
+        let mut hst = {
+            assert!(!hst.is_null());
+            Box::from_raw(hst as *mut History)
+        };
+        let result = hst.prev();
+        let result = CString::new(result).unwrap().into_raw();
+
+        // Do not deallocate the object - C still has a pointer to it
+        mem::forget(hst);
+
+        result
+    })
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_history_next(hst: *mut c_void) -> *mut c_char {
+    abort_on_panic(|| {
+        let mut hst = {
+            assert!(!hst.is_null());
+            Box::from_raw(hst as *mut History)
+        };
+        let result = hst.next();
+        let result = CString::new(result).unwrap().into_raw();
+
+        // Do not deallocate the object - C still has a pointer to it
+        mem::forget(hst);
+
+        result
+    })
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_history_load_from_file(hst: *mut c_void, file: *const c_char) {
+    abort_on_panic(|| {
+        let mut hst = {
+            assert!(!hst.is_null());
+            Box::from_raw(hst as *mut History)
+        };
+        let file = {
+            assert!(!file.is_null());
+            CStr::from_ptr(file)
+        }
+        .to_string_lossy()
+        .into_owned();
+
+        hst.load_from_file(file).unwrap();
```

Comment:
I believe the original C++ code did nothing if it failed to open the file; please preserve this behaviour by explicitly ignoring the return value of `load_from_file`.

Ditto for `save_to_file` call, below.

Reply:
I went for the underscore convention vs simply calling the `ok` function on the result.